docs/architecture/adr-082-socket-protocol.md

```diff
@@ -0,0 +1,259 @@
+  (see [Note 1](#note1)).  In practice, however, this greatly complicates
+  debugging a faulty node, since the only way to respond to errors is to panic
+  the node which loses valuable context that could have been logged.
+
+- There are subtle concurrency management dependencies between the client and
+  the server that are not clearly documented anywhere, and it is very easy for
+  small changes in both the client and the server to lead to tricky deadlocks,
+  panics, race conditions, and slowdowns. As a recent example of this, see
+  https://github.com/tendermint/tendermint/pull/8581.
+
+These limitations are fixable, but one important question is whether it is
+worthwhile to fix them.  We can add request and method identifiers, for
+example, but doing so would be a breaking change to the protocol requiring
+every application using it to update.  If applications have to migrate anyway,
+the stability and language coverage of gRPC have improved a lot, and today it
+is probably simpler to set up and maintain an application using gRPC transport
+than to reimplement the Tendermint socket protocol.
+
+Moreover, gRPC addresses all the above issues out-of-the-box, and requires
+(much) less custom code for both the server (i.e., the application) and the
+client. The project is well-funded and widely-used, which makes it a safe bet
+for a dependency.
+
+## Decision
+
+There is a set of related alternatives to consider:
+
+- Question 1: Designate a single IPC standard for out-of-process applications?
+
+  Claim: We should converge on one (and only one) IPC option for out-of-process
+  applications. We should choose an option that, after a suitable period of
+  deprecation for alternatives, will address most or all the highest-impact
+  uses of Tendermint.  Maintaining multiple options increases the surface area
+  for bugs and vulnerabilities, and we should not have multiple options for
+  basic interfaces without a clear and well-documented reason.
+
+- Question 2a: Choose gRPC and deprecate/remove the socket protocol?
+
+  Claim: Maintaining and improving a custom RPC protocol is a substantial
+  project and not directly relevant to the requirements of consensus. We would
+  be better served by depending on a well-maintained open-source library like
+  gRPC.
+
+- Question 2b: Improve the socket protocol and deprecate/remove gRPC?
+
+  Claim: If we find meaningful advantages to maintaining our own custom RPC
+  protocol in Tendermint, we should treat it as a first-class project within
+  the core and invest in making it good enough that we do not require other
+  options.
+
+**One important consideration** when discussing these questions is that _any
+outcome which includes keeping the socket protocol will have eventual migration
+impacts for out-of-process applications_ regardless. To fix the limitations of
+the socket protocol as it is currently designed will require making _breaking
```

Comment:
Is it true that, for the 3 issues outlined in the problem statement, (1) may be fixed by ABCI++, (2) will require breaking changes to allow errors to propogate properly, and (3) will not be breaking ?